### ISSUE TITLE
Truncation issue with GZIPInputStream when reading bgzip-compressed streams

### ISSUE THREAD

**Issue by jkmatila** [samtools/htsjdk#1691]
### Description of the issue

When reading bgzip-compressed data from a stream (e.g. over the network or from a pipe), htsjdk can stop early and not read the stream all the way to the end, processing partial data. This is due to using `java.util.zip.GZIPInputStream` which has a known bug [JDK-7036144](https://bugs.java.com/bugdatabase/view_bug.do?bug_id=7036144#) affecting reading concatenated gzip streams when the complete size of the stream is not known beforehand.

The problem has been discussed earlier in [the comment thread to broadinstitute/gatk issue #4224](https://github.com/broadinstitute/gatk/issues/4224#issuecomment-360282461), but it seems that a part of the recommended fix, replacing the use of the buggy `GZIPInputStream` class, was not adopted. There also is a comment in [SamInputResource.java](https://github.com/samtools/htsjdk/blob/4.0.2/src/main/java/htsjdk/samtools/SamInputResource.java#L95-L100) which references this problem and the aforementioned JDK bug.

For example, `GzipCompressorInputStream` from Apache Commons Compress would be an alternative which does not suffer from this problem, as noted in [its Javadoc](https://commons.apache.org/proper/commons-compress/apidocs/org/apache/commons/compress/compressors/gzip/GzipCompressorInputStream.html).

### Your environment

* version of htsjdk: 4.0.2
* version of Java: OpenJDK 17.0.8.1 (Eclipse Temurin-17.0.8.1+1), OpenJDK 21 (Eclipse Temurin-21+35)
* which OS: Linux x86_64 (Ubuntu 22.04.3)

### Steps to reproduce

Here is an example program which demonstrates one instance of the problem, `VCFIteratorBuilder` using the problematic `GZIPInputStream` class when reading a bgzipped VCF stream, and only part of the stream gets read. (Note: The code attempts to read a 200 MB file from the Web, although it in practice due to the bug it does not get very far.)

```java
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;

import htsjdk.variant.vcf.VCFIterator;
import htsjdk.variant.vcf.VCFIteratorBuilder;

public class Repro {

    public static void main(String[] args) throws IOException {
        URL url = new URL("https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/release/20130502/ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz");
        int expectedNumRecords = 1103547;

        int numRecords = 0;
        try (InputStream in = url.openStream();
             VCFIterator vcfIterator = new VCFIteratorBuilder().open(in)) {
            while (vcfIterator.hasNext()) {
                vcfIterator.next();
                numRecords++;
                if (numRecords % 100000 == 0) {
                    System.out.println("Read " + numRecords + " records...");
                }
            }
        }

        if (numRecords != expectedNumRecords) {
            System.err.println("URL: " + url);
            System.err.println("Expected: " + expectedNumRecords + " records");
            System.err.println("Parsed: " + numRecords + " records");
            System.exit(1);
        }
    }

}
```

There are also other places where `GZIPInputStream` is used in htsjdk.

### Expected behaviour

The compressed VCF stream should be read all the way to the end and all records parsed.

### Actual behaviour

The code stops reading the stream at some point, not parsing all records. No error is returned and parsing appears successful but in fact produces only part of the expected data. The exact location where it stops probably depends on how the block boundaries happen to coincide with the buffer size. It appears that if you have a large enough input file, it will likely stop at some point before reaching the end.

On both OpenJDK 17 (Eclipse Temurin-17.0.8.1+1) and on OpenJDK 21 (Eclipse Temurin-21+35) using the official [Eclipse Temurin Docker images](https://hub.docker.com/_/eclipse-temurin) the above code outputs the following:

```
URL: https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/release/20130502/ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz
Expected: 1103547 records
Parsed: 7395 records
```

**Comment by lbergelson** [comment]
Yeah, this bug is awful.  We worked around it in GATK but never really fixed it in htsjdk.  I decided to message the java development mailing list last week about it.  I wish I had done that years ago, since someone responded really quickly.  After 12 years or so since it was first reported, this bug is going to be fixed soon.    I'm hoping/asking that it be back ported to java 17 or at least 21.

See https://github.com/openjdk/jdk/pull/17113

**Comment by jrobinso** [comment]
Update, the JDK 23 fix was rolled back.  Accoring to Claude:
```
 JDK 23 (JDK-7036144) stopped calling available() after each gzip member and instead always tried to read the next member's header. That fixed our bug, but it caused two regressions elsewhere:
  - A slowdown (JDK-8374644), which was fixed separately.
  - Stalls on open network connections (JDK-8377896). When a stream holds a single gzip member on a connection that stays open, the extra read waits for the socket's full read timeout. The reporter saw this first in Temurin 21.0.9, the update that
    received the backport.

  The rollback proposal says the second problem "isn't easy to fix." It also says the regressions showed "the usage of InputStream.available() in GZIPInputStream was serving a purpose and cannot be trivially avoided." The JDK team concluded it was
  "better and safer to rollback."

...

  - JDK 21 is unsettled. Its revert backport isn't listed yet. The property's tracker entry (CSR JDK-8386102) lists 21-pool and 17-pool, which means not yet scheduled for a release. A future 21 update may well bring the bug back.
  - Versions 26.0.3+ and 27.x aren't covered. I checked only what the tracker lists, not every release after those. 
  
```

**Comment by jrobinso** [comment]
In IGV we've worked around this by wrapping the GzipInputStream.  I just updated this to handle files decoded in the htsjdk using the URL helper hook,  the commit is here:  https://github.com/igvteam/igv/commit/c41705a593388676719b6cf868d5f4df973ad7c5.   You might be able to do something like this in the htsjdk.

**Comment by jkmatila** [comment]
Given that the JDK fix has been rolled back, I think it would make sense to fix this inside htsjdk. Even if this issue gets a fix in some future JDK, users are going to run htsjdk on old JDK versions for a long time and hit this insidious silent truncation issue. `GzipCompressorInputStream` from Apache Commons Compress with its concatenated-stream support would be seemingly a drop-in-replacement for `java.util.zip.GZIPInputStream` without suffering from this problem. Apache Commons Compress seems to already be a dependency, so this would not even require adding a new dependency.

**Comment by tfenne** [comment]
Thanks @jrobinso and @jkmatila for re-surfacing this.  It's now fixed on master and will go out with the 6.0.0 release.  That's going to be a big release, so it's probably still 2+ weeks away.
